core/src/token_info/cached.rs

```diff
@@ -0,0 +1,159 @@
+use super::{TokenBaseInfo, TokenId, TokenInfoFetching};
+
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt};
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+/**
+ * Implementation of TokenInfoFetching that stores previously fetched information in an in-memory cache for fast retrieval.
+ * TokenIds will always be fetched from the inner layer, as new tokens could be added at any time.
+ */
+struct TokenInfoCache {
+    cache: RwLock<HashMap<TokenId, TokenBaseInfo>>,
```

Comment:
As @e00E mentioned, this should definitely use `async-std` synchronization primitives. This is especially true because of the potential for deadlocks here (seeing as it is possible for futures to run all on a single thread, which may be blocked waiting on a lock, preventing the future holding the lock to progress because the thread is not available for polling it).

Reply:
Makes perfect sense and will change. Just to understand the concept of async a bit better would it really be possible to run into a deadlock here? We are not yielding while we hold the lock, so I don't really see how we can deadlock. Again this is a detail we should not rely on but for educational purposes, how could we deadlock in the current code?

Reply:
Looking at the code closely, I think you are right and I don't see how the current code would deadlock. For some reason I thought that the write handle was held for the whole `async { ... }.boxed()` scope, but it is only held for the `if` statement scope, and as you mentioned we never yield the future while it is held.